src/modalities/checkpointing/stateful/app_state.py

```diff
@@ -0,0 +1,241 @@
+            raise RuntimeError(
+                "Cannot call load_state_dict twice on the same AppState object. " "State dict has already been loaded."
+            )
+
+        ModelStateRetriever.load_state_dict_(app_state=self, state_dict=state_dict[StatefulComponents.MODEL.value])
+        OptimizerStateRetriever.load_state_dict_(
+            app_state=self,
+            state_dict=state_dict[StatefulComponents.OPTIMIZER.value],
+        )
+        if self._lr_scheduler is not None:
+            LRSchedulerStateRetriever.load_state_dict_(
+                app_state=self, state_dict=state_dict[StatefulComponents.LR_SCHEDULER.value]
+            )
+        self._is_loaded = True
+
+
+class StateRetrieverIF(ABC):
+    """State retriever interface for loading and getting state dicts of
+    models, optimizers and lr schedulers. Other stateful components can be added as needed
+    by having the retriever implement this interface.
+    """
+
+    @staticmethod
+    @abstractmethod
+    def load_state_dict_(app_state: AppState, state_dict: dict[str, Any]) -> None:
+        """Loads the state dict into the AppState object.
+
+        Args:
+            app_state (AppState): The application state object.
+            state_dict (dict[str, Any]): The state dict to load into the AppState object.
+
+        Raises:
+            NotImplementedError: This abstract method is not implemented and should be overridden in a subclass.
+        """
+        raise NotImplementedError
+
+    @staticmethod
+    @abstractmethod
+    def get_state_dict(app_state: AppState) -> dict[str, Any]:
+        """Returns the state dict of the AppState object.
+
+        Args:
+            app_state (AppState): The application state object.
+
+        Raises:
+            NotImplementedError: This abstract method is not implemented and should be overridden in a subclass.
+
+        Returns:
+            dict[str, Any]: The state dict of the AppState object.
+        """
+        raise NotImplementedError
+
+
+class ModelStateRetriever(StateRetrieverIF):
+    @staticmethod
+    def get_state_dict(app_state: AppState) -> dict[str, Any]:
+        """Returns the state dict of the model in the AppState object.
+
+        Args:
+            app_state (AppState): The app_state object containing the model.
+
+        Returns:
+            dict[str, Any]: The state dict of the model in the AppState object.
+        """
+        return get_model_state_dict(model=app_state.model)
+
+    @staticmethod
+    def load_state_dict_(app_state: AppState, state_dict: dict[str, Any]) -> None:
+        """Loads the state dict into the model in the AppState object.
+
+        Args:
+            app_state (AppState): The app_state object containing the model.
+            state_dict (dict[str, Any]): The state dict to load into the model.
+        """
+        set_model_state_dict(model=app_state.model, model_state_dict=state_dict, options=StateDictOptions(strict=False))
```

Comment:
Why is strict per default equal to False? Please add a comment.

Reply:
TorchTitan also sets this to False: https://github.com/pytorch/torchtitan/blob/30b9ea0b1ad893379d2ff3b12dbf18600730c249/torchtitan/components/checkpoint.py#L63

However, I'm not entirely sure why. 

There is also not really any documentation on it:
https://docs.pytorch.org/docs/stable/distributed.checkpoint.html#torch.distributed.checkpoint.state_dict.StateDictOptions